### PR TITLE
Convention payment via the API and tests

### PR DESCRIPTION
**Pull request by mbayopanda**
This PR add : 
- A convention API for payment in the route : '/cash/conventions' 
- Add some tests for convention API

### CODE REVIEW

**Comment by mbayopanda** [comment]
@jniles when you'll have the time, could you look this pull-request

File: server/test/api/conventions.js [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+  /** login before each request */
+  beforeEach(helpers.login(agent));
+
+  /** Test with dates */
+  var date = new Date();
+  var futurDate = new Date(date.getYear() + 1900, date.getMonth(), date.getDay() + 10);
+
+  /** Primary Cash Object */
+  var primaryCash = {
+      uuid        : uuid(),
```
Comment (jniles) [inline server/test/api/conventions.js]: This usage of `node-uuid` is deprecated.  [See this notice](https://www.npmjs.com/package/node-uuid#deprecated-apis).  The correct usage of `node-uuid` is like this:

``` js
var uuid = require('node-uuid');
var primaryCash = {
  uuid : uuid.v4(),
  /* ... other properties ... */
};
```

File: server/test/api/conventions.js [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  'use strict';
+
+  var agent = chai.request.agent(helpers.baseUrl);
+
+  /** login before each request */
+  beforeEach(helpers.login(agent));
+
+  /** Test with dates */
+  var date = new Date();
+  var futurDate = new Date(date.getYear() + 1900, date.getMonth(), date.getDay() + 10);
```
Comment (jniles) [inline server/test/api/conventions.js:31]: This is kind of random -- is there a reason behind these choices of numbers?

File: server/controllers/finance/conventions.js [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    });
+  }
+
+  /** Get Primary Cash Items */
+  function getPrimaryCashItems(maxAmount, primaryCashUuid) {
+    var items = [];
+    var costReceived = maxAmount;
+
+    /** Enterprise currency is dollars */
+    var isDollars = qData.currency_id === req.session.enterprise.currency_id;
```
Comment (jniles) [inline server/controllers/finance/conventions.js]: A better name is `isEnterpriseCurrency` -- not all enterprises will use dollars!

File: server/controllers/finance/conventions.js [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+        'WHERE DATE(e.date) = DATE(?) ';
+    return db.exec(query, [qData.date]);
+  }
+
+  function handleExchangeRate(rates) {
+    if (!rates.length) { throw new req.codes.ERR_EXCHANGE_RATE_NOT_FOUND(); }
+    exchangeRate = rates[0];
+  }
+
+  function writePay() {
```
Comment (jniles) [inline server/controllers/finance/conventions.js:103]: This is a very nicely commented and laid out function.  Thanks!

File: server/controllers/finance/conventions.js [hunk 5]
```diff
@@ -0,0 +1,163 @@
...
+  .then(getUnpayedInvoices)
+  .then(getExchangeRate)
+  .then(handleExchangeRate)
+  .then(writePay)
+  .then(writeToJournal)
+  .catch(next)
+  .done();
+
+  function getConventionUuid(rows) {
+    /** return data to debtorGroups.fetchInvoices(id, balanced, codes) */
```
Comment (jniles) [inline server/controllers/finance/conventions.js]: I do not understand this comment.  Where is `debtorGroups.fetchInvoices()`?

File: server/controllers/finance/conventions.js [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    var isDollars = qData.currency_id === req.session.enterprise.currency_id;
+
+    /** Sort receipts by date ASC */
+    overviews.sort(function (a, b) {
+      return a.trans_date > b.trans_date;
+    });
+
+    /** Collecting primary cash items */
+    var row, balance, allocatedCost;
+    for (var i = 0; i < overviews.length; i++) {
```
Comment (jniles) [inline server/controllers/finance/conventions.js:149]: This is a really nice algorithm.  It is very readable and the variables are clearly named.

**Comment by jniles** [comment]
@mbayopanda, this is a good PR.  If the tests pass, I'm happy to have this merged in.  I've made a few comments that you should look at -- none of them will block this PR, but if you want to fix them, let me know and I'll wait until you are done.

**Update** This PR broke the accounts API tests.  It looks like you added an account or two.  Could you update those tests?

**Comment by jniles** [comment]
:+1: LGTM.

**Pull request merged**